pg_shard--1.2.sql

```diff
@@ -3,57 +3,172 @@
-	'pgs_distribution_metadata.shard_placement', '');
-SELECT pg_catalog.pg_extension_config_dump(
-	'pgs_distribution_metadata.partition', '');
+-- needed in our views
+CREATE FUNCTION column_to_column_name(table_oid oid, column_var text)
+RETURNS text
+AS 'MODULE_PATHNAME'
+LANGUAGE C STABLE STRICT;
+
+DO $$
+DECLARE
+	use_citus_metadata boolean := false;
+BEGIN
+	BEGIN
+		PERFORM 'pg_catalog.pg_dist_partition'::regclass;
+		use_citus_metadata = true;
+	EXCEPTION
+		WHEN undefined_table THEN
+			use_citus_metadata = false;
+	END;
+
+	IF use_citus_metadata THEN
+		CREATE FUNCTION adapt_and_insert_shard() RETURNS TRIGGER AS $aais$
+		BEGIN
+			IF NEW.id IS NULL THEN
+				NEW.id = nextval('pg_dist_shardid_seq');
+			END IF;
+
+			INSERT INTO pg_dist_shard
+						(logicalrelid,
+						 shardid,
+						 shardstorage,
+						 shardalias,
+						 shardminvalue,
+						 shardmaxvalue)
+			VALUES      (NEW.relation_id,
+						 NEW.id,
+						 NEW.storage,
+						 NULL,
+						 NEW.min_value,
+						 NEW.max_value);
+
+			RETURN NEW;
+		END
+		$aais$ LANGUAGE plpgsql;
+
+		CREATE FUNCTION adapt_and_insert_shard_placement() RETURNS trigger AS $aaisp$
+		BEGIN
+			INSERT INTO pg_dist_shard_placement
+						(shardid,
+						 shardstate,
+						 shardlength,
+						 nodename,
+						 nodeport)
+			VALUES      (NEW.shard_id,
+						 NEW.shard_state,
+						 0,
+						 NEW.node_name,
+						 NEW.node_port)
+			RETURNING oid INTO STRICT NEW.id;
+
+			RETURN NEW;
+		END
+		$aaisp$ LANGUAGE plpgsql;
+
+		CREATE FUNCTION adapt_and_insert_partition() RETURNS trigger AS $aaip$
+		BEGIN
+			INSERT INTO pg_dist_partition
+						(logicalrelid,
+						 partmethod,
+						 partkey)
+			VALUES      (NEW.relation_id,
+						 NEW.partition_method,
+						 column_name_to_column(NEW.relation_id, NEW.key));
+
+			RETURN NEW;
+		END
+		$aaip$ LANGUAGE plpgsql;
+
+		-- metadata relations are views under CitusDB
+		CREATE SCHEMA pgs_distribution_metadata
+			CREATE VIEW shard AS
+				SELECT shardid       AS id,
+					   logicalrelid  AS relation_id,
+					   shardstorage  AS storage,
+					   shardminvalue AS min_value,
+					   shardmaxvalue AS max_value
+				FROM   pg_dist_shard
+
+			CREATE TRIGGER shard_insert INSTEAD OF INSERT ON shard
+				FOR EACH ROW
+				EXECUTE PROCEDURE adapt_and_insert_shard()
+
+			CREATE VIEW shard_placement AS
+				SELECT oid::bigint AS id,
+					   shardid     AS shard_id,
+					   shardstate  AS shard_state,
+					   nodename    AS node_name,
+					   nodeport    AS node_port
+				FROM   pg_dist_shard_placement
+
+			CREATE TRIGGER shard_placement_insert INSTEAD OF INSERT ON shard_placement
+				FOR EACH ROW
+				EXECUTE PROCEDURE adapt_and_insert_shard_placement()
+
+			CREATE VIEW partition AS
+				SELECT logicalrelid AS relation_id,
+					   partmethod   AS partition_method,
+					   column_to_column_name(logicalrelid, partkey) AS key
+				FROM   pg_dist_partition
+
+			CREATE TRIGGER partition_insert INSTEAD OF INSERT ON partition
```

Comment:
We trigger on INSERT's. Will updates/deletes error out? 

If so, do we need some check in the shard-repair functionality that errors if its on CitusDB (with message to use the shard rebalancer maybe)?